### PR TITLE
New gaa API

### PR DESCRIPTION
**Pull request by rew326**

### CODE REVIEW

File: examples/sample-pub/views/article-ea-simplified-api.html [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+        return new Promise ((resolve, reject) => {
+          publlisherEntitlement = {
+            granted: false,
+          };
+
+          resolve(publlisherEntitlement);
+        });
+      }
+      
+      function handleLoginPromise() {
```
Comment (jpettitt) [inline examples/sample-pub/views/article-ea-simplified-api.html:145]: Comment explaining why this is different to get user state.

File: examples/sample-pub/views/article-ea-simplified-api.html [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+      (self.SWG = self.SWG || []).push(function (subscriptions) {
+        // Set up metering demo controls.
+        MeteringDemo.setupControls();
+        // Set up language.
+        MeteringDemo.setupLanguage();
+
+        GaaMetering.init({
+          params: {
+          productId: 'examplenews.com:showcase',
+          googleSignInClientId: '520465458218-e9vp957krfk2r0i4ejeh6aklqm7c25p4.apps.googleusercontent.com',
```
Comment (jpettitt) [inline examples/sample-pub/views/article-ea-simplified-api.html:185]: see comment about constants

File: examples/sample-pub/sample-pub-app.js [hunk 3]
```diff
@@ -334,6 +334,29 @@ app.get('/redirect-to/swg-basic.js', (req, res) => {
...
 
+/**
+ * EA - Simplified API
+ * TODO: make it a flow of "An Article"
+ */
+ app.get('/ea-simplified-api', (req, res) => {
+  const id = 1;
+  const article = ARTICLES[id - 1];
+  const prevId = id - 1 >= 0 ? String(id - 1) : false;
+  const nextId = id + 1 < ARTICLES.length ? String(id + 1) : false;
```
Comment (jpettitt) [inline examples/sample-pub/sample-pub-app.js:345]: Explain how this differs from production usage

**Review by jpettitt — DISMISSED** [review]
see my comments

**Comment by jpettitt** [comment]
Nudge @rew326

File: examples/sample-pub/views/article-ea-simplified-api.html [hunk 4]
```diff
@@ -0,0 +1,318 @@
+<!doctype html>
```
[the remark this answers is not in the document]
Comment (rew326) [reply examples/sample-pub/views/article-ea-simplified-api.html:1]: Added comments

File: examples/sample-pub/views/article-ea-simplified-api.html (hunk 2 above)
Comment (rew326) [reply examples/sample-pub/views/article-ea-simplified-api.html]: Added comments

File: examples/sample-pub/views/article-ea-simplified-api.html [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+      // This is used if you redirect users or refresh the page during the Handle registration
+      // for new users or Handle login for existing users flows.
+      const allowedReferrers = ["example.com","test.com","localhost"]
+
+      // User's Showcase entitlement determined on your server
+      // Optional - server side only
+      const showcaseEntitlement = urlParams.get("showcaseEntitlement");
+
+      // The userState object for the current user.
+      function getUserState() {
```
Comment (jpettitt) [inline examples/sample-pub/views/article-ea-simplified-api.html:113]: Add more context here -either point to doc or explain timestamps usage

**Comment by jpettitt** [comment]
Don't include yarn.lock in PR 

Still more comments - really *every* place there is a constant you need to explain it (eg you have the user state in multiple places withe constant timestamps).  People will copy and past chunks of this code and unless it's super obvious they need to change things they will get it wrong.  I know it may feel like overkill but it's better than people wasting time trying to fix stupid bugs from copy pasting the wrong constant.

In particular some text about how the user state object must change after various events would be helpful.  Also explain what to do with timestamps when they are not applicable (not yet subscribed, not yet registered) - maybe replace all the static user state examples with a call to function that returns a promise of new user state and then in the function document it to death?

File: examples/sample-pub/sample-pub-app.js (hunk 3 above)
Comment (rew326) [reply examples/sample-pub/sample-pub-app.js:345]: Redundant code, removed.

File: src/utils/gaa-test.js [hunk 6]
```diff
@@ -39,10 +40,14 @@ import {
 import {I18N_STRINGS} from '../i18n/strings';
 import {JwtHelper} from './jwt';
 import {tick} from '../../test/tick';
+import {resolve} from 'bluebird';
```
Comment (jpettitt) [inline src/utils/gaa-test.js:43]: Why do you need bluebird here? Node has native promises.

Comment (rew326) [reply src/utils/gaa-test.js]: Removed

File: examples/sample-pub/views/article-ea-simplified-api.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (rew326) [reply examples/sample-pub/views/article-ea-simplified-api.html:1]: Could you please take a look?

File: examples/sample-pub/views/article-ea-simplified-api.html (hunk 2 above)
Comment (rew326) [reply examples/sample-pub/views/article-ea-simplified-api.html]: Could you please take a look?

File: src/utils/gaa.js [hunk 7]
```diff
@@ -1237,3 +1386,659 @@ export class GaaUtils {
...
+      return jsonLdPageConfig;
+    }
+
+    const microdataPageConfig =
+      GaaMetering.isArticleFreeFromMicrodataPageConfig_();
+    if (microdataPageConfig) {
+      return microdataPageConfig;
+    }
+
+    return false;
```
[the remark this answers is not in the document]
Comment (sidneyzanetti) [reply src/utils/gaa.js]: done.

File: src/utils/gaa.js [hunk 8]
```diff
@@ -1237,3 +1386,659 @@ export class GaaUtils {
...
+            debugLog('unlocked for subscriber');
+          } else if (userState.grantReason === 'FREE') {
+            subscriptions.setShowcaseEntitlement({
+              entitlement: ShowcaseEvent.EVENT_SHOWCASE_UNLOCKED_FREE_PAGE,
+              isUserRegistered: GaaMetering.isUserRegistered(userState),
+            });
+            debugLog('unlocked for free');
+          } else if (userState.grantReason === 'METERING') {
+            // The user has access from the publisher's meter
+            subscriptions.setShowcaseEntitlement({
```
[the remark this answers is not in the document]
Comment (sidneyzanetti) [reply src/utils/gaa.js:1591]: done

File: src/utils/gaa.js [hunk 9]
```diff
@@ -1237,3 +1386,659 @@ export class GaaUtils {
...
+    const userState = params.userState;
+    const unlockArticle = params.unlockArticle;
+    const handleSwGEntitlement = params.handleSwGEntitlement;
+
+    const registerUserPromise = params.registerUserPromise;
+    const handleLoginPromise = params.handleLoginPromise;
+    const publisherEntitlementPromise = params.publisherEntitlementPromise;
+
+    // Validate gaa parameters and referrer
+    if (!GaaMetering.isGaa(allowedReferrers)) {
```
[the remark this answers is not in the document]
Comment (rew326) [reply src/utils/gaa.js:1448]: isGaa is not just checking if the domain was among the referrers in params or was from Google but also checking gaa parameters. We thought that this a separate process from just validating parameters but the action of checking if EA should be triggered at all. Do you have a strong preference here?

[the remark this answers is not in the document]
Comment (adolgachev) [reply src/utils/gaa.js:1448]: Fair enough.

File: src/utils/gaa.js [hunk 10]
```diff
@@ -1237,3 +1386,659 @@ export class GaaUtils {
...
+      });
+    }
+
+    function unlockArticleIfGranted() {
+      if (!GaaMetering.validateUserState(userState)) {
+        debugLog('Invalid userState object');
+        return false;
+      } else if (userState.granted === true) {
+        callSwg((subscriptions) => {
+          if (userState.grantReason === 'SUBSCRIBER') {
```
[the remark this answers is not in the document]
Comment (adolgachev) [reply src/utils/gaa.js]: I'm okay with keeping this as-is given that. I think I'd be more inclined to go with it if we were using typescript.

File: src/utils/gaa.js [hunk 11]
```diff
@@ -372,6 +405,31 @@ export let GoogleIdentityV1;
...
+ * handleLoginPromise: (Promise|null),
+ * caslUrl: string,
+ * handleSwGEntitlement: function(): ?,
+ * publisherEntitlementPromise: (Promise|null),
+ * registerUserPromise: (Promise|null),
+ * showPaywall: function(): ?,
+ * showcaseEntitlement: string,
+ * unlockArticle: function(): ?,
+ * userState: {
+ *   grantReason: string,
```
[the remark this answers is not in the document]
Comment (rew326) [reply src/utils/gaa.js:423]: Agreed to keep it as it is to not increase complexity for partners.

**Review by adolgachev — APPROVED** [review]

**Pull request merged**